proposals/rich-content-graphql-api.md

```diff
@@ -0,0 +1,331 @@
+      "diagram.svg",
+      "thumbnail001.jpg",
+      "thumbnail002.jpg",
+      "video.mp4"
+    ],
+    "descendentRoles": ["Banner", "Accordion", "Video", "Image", "Row", "Text"],
+    "children": [
+      {
+        "role": "Banner",
+        "depth": 0,
+        "preload": ["banner1.jpg"],
+        "descendentRoles": []
+      },
+      {
+        "role": "Accordion",
+        "depth": 3,
+        "preload": [
+          "diagram.svg",
+          "thumbnail001.jpg",
+          "thumbnail002.jpg",
+          "video.mp4"
+        ],
+        "descendentRoles": ["Video", "Image", "Row"]
+      },
+      {
+        "role": "Text",
+        "depth": 0,
+        "preload": [],
+        "descendentRoles": []
+      }
+    ],
+    "html": "Hello world! Text has already loaded."
+  }
+}
+```
+
+The PWA would handle this response:
+
+1.  In parallel:
+
+    1.  Render the `html` content by injecting it into a `RawHTML` component.
+
+    2.  Initiate preload requests for all the URLs in the top-level `preload` field. *Note that the responses "roll up" descendent preloads to a topmost list, as well as descendentRoles.*
+
+    3.  In series:
+
+        1.  In parallel, dynamically load and inject the React Components for:
+
+            - `PageBuilder_Column`
+            - `PageBuilder_Banner`
+            - `PageBuilder_Accordion`
+            - `PageBuilder_Video`
+            - `PageBuilder_Image`
+            - `PageBuilder_Row`
+            - `PageBuilder_Text`
+
+            (The bundle system and service worker will load from their caches if possible.)
+
+        2.  Run a subsequent query provided by the components' respective implementations:
+
+            ```graphql
+            fragment PageBuilder_Column_Config on PageBuilder_Column {
+              appearance
+              backgroundColor
+              backgroundImage
+              backgroundPosition
+              backgroundSize
+              backgroundRepeat
+              backgroundAttachment
+              minHeight
+              verticalAlign
+            }
+
+            fragment PageBuilder_Banner_Config on PageBuilder_Banner {
+              appearance
+              backgroundType
+              "More banner config here."
+            }
+
+            """
+            Additional fragments here.
+            """
+
+            content {
+              ...PageBuilder_Column_Config
+              children {
+                ... on PageBuilder_Banner {
+                  ...PageBuilder_Banner_Config
+                }
+                ... on PageBuilder_Accordion {
+            	  ...PageBuilder_Accordion_Config
+                }
+                ... on PageBuilder_Text {
+                  ...PageBuilder_Text_Config
+                }
+            	children {
+                  ...NodeMetadata
+                }
+              }
+            }
+            ```
+
+        3.  Re-render the fully hydrated second level of component when that query returns results.
```

Comment:
Any way to avoid a re-render? I'm concerned about the page jumping as new content loads. I know you're not a fan of SSR, but it sounds like it would solve this problem.

Reply:
In order to avoid this, the HTML fallbacks would have to have dimension. I think that the RichContent component could load a root PageBuilder stylesheet which would apply default dimensions.

As more PageBuilder components become "known" to the PWA runtime, it could add fragments to VisualContentNode queries like `...on PageBuilderGallery {}` etc. which would get more and more specific configuration, including dimension, on the first render...